### PR TITLE
[Hunter] Fix Vulnerable stats secondary target procs

### PR DESCRIPTION
**Pull request by Jayezi**
Was simming Piercing Shot and thought the APL was broken at first.

### CODE REVIEW

File: engine/class_modules/sc_hunter.cpp [hunk 1]
```diff
@@ -730,11 +730,9 @@ struct vulnerability_stats_t
     }
   }
 
-  void update( hunter_t* p, const action_t* a )
+  void update( hunter_t* p, const action_state_t* state )
```
Comment (nuoHep) [inline engine/class_modules/sc_hunter.cpp]: You could name it `s` and avoid most of the churn below.

File: engine/class_modules/sc_hunter.cpp [hunk 2]
```diff
@@ -3531,10 +3529,21 @@ struct piercing_shot_t: public hunter_ranged_attack_t
...
+    if ( execute_state -> target != p() -> target )
+    {
+      original_target = execute_state -> target;
+      execute_state -> target = p() -> target;
+    }
+
+    vulnerability_stats.update( p(), execute_state );
+
+    if ( original_target )
+      execute_state -> target = original_target;
```
Comment (nuoHep) [inline engine/class_modules/sc_hunter.cpp]: This one I'm not sure about. I always though that the target from execute state is the actual target of the action, so modifying it to point to the current player target is actually "backwards". Maybe I'm simply wrong or missing smth.

Comment (nuoHep) [reply engine/class_modules/sc_hunter.cpp]: Yeah... looks like I was wrong this whole time, at least for aoe actions.
But, using player target is also wrong, and I believe we should be using action -> target here.

**Comment by nuoHep** [comment]
2 comments, the second one I'm not sure about.
One thing worth checking is if using `s -> action -> target` inside the update method (and dropping the whole hunk inside pis execute) will give the same results (I believe it should).
Or simply using `a -> target` and leaving update signature the same.

**Comment by nuoHep** [comment]
Thanks

**Comment by navv1234** [comment]
I'm unsure what all this is, but generally speaking, action_state_t::target is always the safe bet for the actual target of the ability.

action_t::target is not the correct solution in all situations.

**Comment by Jayezi** [comment]
For piercing shot the state's target had been looped to a secondary target by the time execute returned but I wanted to track specific info on the primary target (it does more damage to primary/primary is what we're trying to maximize the patient sniper stacks on) and other info on the secondary targets. Based on the apl the primary target is always the player target but yea potentially the action target could be different, but at that point something else is going on or something happened to the apl since all the vulnerable window stuff it uses is based off the current target and done before it gets to actions.

**Comment by navv1234** [comment]
Yeah we went through this with @nuoHep in discord, it's all good.

**Pull request merged**